cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,236 @@
+   * as caching it.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch
+   * @return The Local Location for this artifact
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching or caching the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public Location getArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Fetching artifact info for {}", artifactId);
+    Location localLocation = getArtifactDirLocation(artifactId);
+    Long lastModifiedTimestamp = null;
+    // If the local cache exists, check if we have a timestamp directory
+    if (localLocation.exists()) {
+      String[] fileList = Paths.get(localLocation.toURI()).toFile().list();
+      if (fileList != null && fileList.length > 0) {
+        //Split on period to get timestamp without the .jar extension
+        lastModifiedTimestamp = Arrays.stream(fileList).map(s -> Long.valueOf(s.split("\\.")[0]))
+          .max(Long::compare).get();
+      }
+    }
+
+    String namespaceId = artifactId.getNamespace();
+    ArtifactScope scope = ArtifactScope.USER;
+    // Cant use 'system' as the namespace in the request because that generates an error, the namespace doesnt matter
+    // as long as it exists. Using default because it will always be there
+    if (ArtifactScope.SYSTEM.toString().equalsIgnoreCase(namespaceId)) {
+      namespaceId = "default";
+      scope = ArtifactScope.SYSTEM;
+    }
+
+    String url = String.format("namespaces/%s/artifacts/%s/versions/%s/download?scope=%s",
+                               namespaceId,
+                               artifactId.getArtifact(),
+                               artifactId.getVersion(),
+                               scope);
+
+    if (lastModifiedTimestamp != null) {
+      LOG.debug("Found existing local version with timestamp {}", lastModifiedTimestamp);
+      url += String.format("&lastModified=%d", lastModifiedTimestamp);
+    }
+
+    Long finalLastModifiedTimestamp = lastModifiedTimestamp;
+    String finalUrl = url;
+    Location newJarLocation = Retries
+      .callWithRetries(() -> fetchArtifact(artifactId, finalUrl, finalLastModifiedTimestamp), retryStrategy);
+
+    // If we didn't have a version that was previous cached then we're done and we can return the location
+    if (lastModifiedTimestamp == null) {
+      return newJarLocation;
+    }
+
+    // This means we already have a jar but its out of date, we should delete the jar and the unpacked directory
+    Location oldJarLocation = getArtifactJarLocation(artifactId, lastModifiedTimestamp);
+    LOG.debug("Got new location as {} with old location as {}", newJarLocation, oldJarLocation);
+    if (!newJarLocation.equals(oldJarLocation)) {
+      Location oldUnpackLocation = getUnpackLocalPath(artifactId, lastModifiedTimestamp);
```

Comment:
Instead of doing the deletion in-place, it might be better to have a service to do cleanup of outdated files periodically.

Reply:
Do you think this is required or a nice-to-have that can be added in a followup PR?

Reply:
Created [CDAP-18051](https://cdap.atlassian.net/browse/CDAP-18051) to track this work. Will submit it as a followup PR.